subsys/net/lib/wifi_mgmt_ext/wifi_mgmt_ext.c

```diff
@@ -0,0 +1,228 @@
+#include "common/defs.h"
+#include "wpa_supplicant/config.h"
+#include "wpa_supplicant_i.h"
+#include "driver_i.h"
+
+extern struct k_sem wpa_supplicant_ready_sem;
+extern struct k_mutex wpa_supplicant_mutex;
+extern struct wpa_global *global;
+
+
+LOG_MODULE_REGISTER(wifi_mgmt_ext, CONFIG_WIFI_MGMT_EXT_LOG_LEVEL);
+
+/* helper functions {*/
+
+static inline struct wpa_supplicant *get_wpa_s_handle(const struct device *dev)
+{
+	struct wpa_supplicant *wpa_s = NULL;
+	int ret = k_sem_take(&wpa_supplicant_ready_sem, K_SECONDS(2));
+
+	if (ret) {
+		LOG_ERR("%s: WPA supplicant not ready: %d", __func__, ret);
+		return NULL;
+	}
+
+	k_sem_give(&wpa_supplicant_ready_sem);
+
+	wpa_s = wpa_supplicant_get_iface(global, dev->name);
+	if (!wpa_s) {
+		LOG_ERR("%s: Unable to get wpa_s handle for %s\n", __func__, dev->name);
+		return NULL;
+	}
+
+	return wpa_s;
+}
+
+static inline int channel_to_freq(const struct wifi_connect_req_params *params,
+					   struct wpa_ssid *ssid)
+{
+	if (params->channel == WIFI_CHANNEL_ANY) {
+		return 0;
+	}
+
+	/* We use a global channel list here and also use the widest
+	 * op_class for 5GHz channels as there is no user input
+	 * for these.
+	 */
+	int freq  = ieee80211_chan_to_freq(NULL, 81, params->channel);
+
+	if (freq <= 0) {
+		freq  = ieee80211_chan_to_freq(NULL, 128, params->channel);
+	}
+
+	if (freq <= 0) {
+		LOG_ERR("Invalid channel %d", params->channel);
+		return -EINVAL;
+	}
+
+	ssid->scan_freq = os_zalloc(2 * sizeof(int));
+	if (!ssid->scan_freq) {
+		return -ENOMEM;
+	}
+	ssid->scan_freq[0] = freq;
+	ssid->scan_freq[1] = 0;
+
+	ssid->freq_list = os_zalloc(2 * sizeof(int));
+	if (!ssid->freq_list) {
+		os_free(ssid->scan_freq);
+		return -ENOMEM;
+	}
+	ssid->freq_list[0] = freq;
+	ssid->freq_list[1] = 0;
+
+	return 0;
+}
+
+static inline int copy_psk(const struct wifi_connect_req_params *params,
+				    struct wpa_ssid *ssid)
+{
+	if (!params->psk && !params->sae_password) {
+		return 0;
+	}
+	if (params->security == 3) {
+		ssid->key_mgmt = WPA_KEY_MGMT_SAE;
+		str_clear_free(ssid->sae_password);
+		ssid->sae_password = dup_binstr(params->sae_password, params->sae_password_length);
+
+		if (ssid->sae_password == NULL) {
+			LOG_ERR("%s:Failed to copy sae_password\n", __func__);
+			return -ENOMEM;
+		}
+	} else {
+		if (params->security == 2)
+			ssid->key_mgmt = WPA_KEY_MGMT_PSK_SHA256;
+		else
+			ssid->key_mgmt = WPA_KEY_MGMT_PSK;
+
+		str_clear_free(ssid->passphrase);
+		ssid->passphrase = dup_binstr(params->psk, params->psk_length);
+
+		if (ssid->passphrase == NULL) {
+			LOG_ERR("%s:Failed to copy passphrase\n", __func__);
+			return -ENOMEM;
+		}
+	}
+
+	wpa_config_update_psk(ssid);
+
+	return 0;
+}
+
+static inline int add_network(struct wpa_supplicant *wpa_s, struct wifi_connect_req_params *params)
+{
+	int ret = 0;
+	struct wpa_ssid *network = wpa_supplicant_add_network(wpa_s);
+
+	if (network == NULL) {
+		return -EINVAL;
+	}
+
+	network->ssid = os_zalloc(sizeof(u8) * WIFI_SSID_MAX_LEN);
```

Comment:
Do I understand correctly, that wpa_supplicant is responsible for freeing the memory, hence the need of using the allocator from the supplicant?

Reply:
Yes, this is freed either by remove_network or wpa_supplicant de-init.